File: src/main/java/com/example/finsentinel/service/trading/engine/TradingEngine.java
package com.example.finsentinel.service.trading.engine;

import java.util.List;

/**
 * Provider-agnostic trading engine interface (OpenAlice pattern).
 * Implementations: PaperTradingEngine (simulated), AlpacaTradingEngine (US equities),
 * CcxtTradingEngine (crypto).
 */
public interface TradingEngine {
    OrderResult placeOrder(OrderRequest request);
    List<PositionInfo> getPositions();
    List<OrderResult> getOrders();
    AccountInfo getAccount();
    boolean cancelOrder(String orderId);
    String engineName();

    /**
     * Polls the broker for the latest status of open/pending orders.
     * Returns a list of orders whose status has changed since last check.
     * Default implementation returns the full order list (no diffing).
     */
    default List<OrderResult> syncOrders() {
        return getOrders();
    }
}

File: apps/api/src/trading/interfaces/trading-engine.ts
import type {
  OrderRequest,
  OrderResult,
  PositionInfo,
  AccountInfo,
  MarketClock,
} from './types';

/**
 * Contract for trading engines (Paper, Alpaca, OKX, etc.).
 *
 * Each engine handles order placement, position tracking, and account info
 * for a specific broker or simulation mode.
 *
 * Shared contract implemented by all trading engines.
 */
export interface TradingEngine {
  /** Place a buy or sell order. */
  placeOrder(request: OrderRequest): Promise<OrderResult>;

  /** Get all current positions enriched with live market prices. */
  getPositions(): Promise<PositionInfo[]>;

  /** Get order history. */
  getOrders(): Promise<OrderResult[]>;

  /** Get account summary (cash, total value, buying power). */
  getAccount(): Promise<AccountInfo>;

  /** Cancel an order by ID. Returns true if cancelled, false otherwise. */
  cancelOrder(orderId: string): Promise<boolean>;

  /** Unique engine identifier, e.g. "paper", "alpaca", "okx". */
  engineName(): string;

  /**
   * Sync orders with the broker.
   * Default implementation returns getOrders() (no-op sync).
   */
  syncOrders?(): Promise<OrderResult[]>;

  /**
   * Get market clock info.
   * Default: always open.
   */
  getMarketClock?(): Promise<MarketClock>;
}
